lib/Service/EncryptService.php

```diff
@@ -0,0 +1,399 @@
+	 *
+	 * @var array
+	 */
+	protected $supportedAlgos = array(
+		'aes-256-cbc' => array('name' => 'AES-256', 'keySize' => 32, 'blockSize' => 32),
+		'bf' => array('name' => 'BF', 'keySize' => 16, 'blockSize' => 8),
+		'des' => array('name' => 'DES', 'keySize' => 7, 'blockSize' => 8),
+		'des-ede3' => array('name' => 'DES-EDE3', 'keySize' => 21, 'blockSize' => 8), // 3 different 56-bit keys
+		'cast5' => array('name' => 'CAST5', 'keySize' => 16, 'blockSize' => 8),
+	);
+
+	/**
+	 * Supported encryption modes
+	 *
+	 * @var array
+	 */
+	protected $supportedModes = array(
+		'cbc' => 'CBC',
+	);
+
+	public $encrypted_credential_fields = array(
+		'description', 'username', 'password', 'files', 'custom_fields', 'otp', 'email', 'tags', 'url'
+	);
+
+
+	private $server_key;
+	/**
+	 * A class to handle secure encryption and decryption of arbitrary data
+	 *
+	 *  Note that this is not just straight encryption. It also has a few other
+	 *  features in it to make the encrypted data far more secure.  Note that any
+	 *  other implementations used to decrypt data will have to do the same exact
+	 *  operations.
+	 *
+	 * Security Benefits:
+	 *
+	 * - Uses Key stretching
+	 * - Hides the Initialization Vector
+	 * - Does HMAC verification of source data
+	 *
+	 */
+
+
+	/**
+	 * @var string $cipher The openssl cipher to use for this instance
+	 */
+	protected $cipher = '';
+
+	/**
+	 * @var int $rounds The number of rounds to feed into PBKDF2 for key generation
+	 */
+	protected $rounds = 100;
+
+	/**
+	 * Constructor!
+	 *
+	 * @param SettingsService $settings
+	 */
+	public function __construct(SettingsService $settings) {
+		$this->cipher = $settings->getAppSetting('server_side_encryption');
+		$this->rounds = (int)100;
+		$this->server_key = \OC::$server->getConfig()->getSystemValue('passwordsalt', '');
+	}
+
+	/**
+	 * Create an encryption key. Based on given parameters
+	 *
+	 * @param string $userKey The user key to use. This should be specific to this user.
+	 * @param string $serverKey The server key
+	 * @param string $userSuppliedKey A key from the credential (eg guid, name or tags)
+	 * @return string
+	 */
+
+	public static function makeKey($userKey, $serverKey, $userSuppliedKey) {
+		$key = hash_hmac('sha512', $userKey, $serverKey);
+		$key = hash_hmac('sha512', $key, $userSuppliedKey);
+		return $key;
+	}
+
+	/**
+	 * Get the maximum key size for the selected cipher and mode of operation
+	 *
+	 * @return int Value is in bytes
+	 */
+	public function getKeySize() {
+		return $this->supportedAlgos[$this->cipher]['keySize'];
+	}
+
+	/**
+	 * Decrypt the data with the provided key
+	 *
+	 * @param string $data_hex The encrypted datat to decrypt
+	 * @param string $key The key to use for decryption
+	 *
+	 * @returns string|false The returned string if decryption is successful
+	 *                           false if it is not
+	 */
+	public function decrypt($data_hex, $key) {
+
+		if (!function_exists('hex2bin')) {
```

Comment:
Ah good point. We'll need to create a `Polyfills` class that we can import.
But that's kina out of the scope for this PR imo.